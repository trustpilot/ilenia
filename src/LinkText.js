import React from 'react';
import PropTypes from 'prop-types';
import { parse } from 'htmlstring-to-react';
import { interpolate } from './';
import { useTranslations } from './';

const escapeRegex = (str) => str.replace(/([.?*+^$[\]\\(){}|-])/g, '\\$1');

const linkRegex = (startTag, endTag) =>
  new RegExp(`${escapeRegex(startTag)}(.*)${escapeRegex(endTag)}`, 'g');

const addDefaultValues = (link) => ({
  ...link,
  start: link.start || '[LINK-BEGIN]',
  end: link.end || '[LINK-END]',
});

const LinkText = ({ id, links, interpolations, tag }) => {
  const [translations] = useTranslations();
  if (!translations[id]) {
    console.error(`Couldn't find '${id}' in the translations table`); // eslint-disable-line no-console
    return <span />;
  }

  let translated = translations[id];
  const overrides = {};

  links.map(addDefaultValues).map((link, index) => {
    const regexp = linkRegex(link.start, link.end);
    const key = `link-text-${index}`;

    const props = {
      className: link.class || link.className,
      href: link.href,
      id: link.id,
      key: key,
      target: link.target,
      onClick: link.onClick,
    };

    Object.keys(props).forEach((key) => typeof props[key] === 'undefined' && delete props[key]);

    overrides[`a[key="${key}"]`] = (_, textContent) => (
      <a {...props}>
        {interpolations ? interpolate(textContent, interpolations, tag) : textContent}
      </a>
    );

    function insertTag(match) {
      const textContent = match.substring(link.start.length, match.length - link.end.length);
      return `<a key="${key}">${textContent}</a>`;
    }

    translated = translated.replace(regexp, insertTag);
    return translated;
  });

  return parse(translated, {
    dom: {
      ADD_ATTR: ['target', 'key'],
    },
    overrides,
  }).reduce(
    (arr, item) =>
      arr.concat(
        typeof item === 'string' && interpolations ? interpolate(item, interpolations, tag) : [item]
      ),
    []
  );
};

LinkText.propTypes = {
  id: PropTypes.string.isRequired,
  links: PropTypes.arrayOf(
    PropTypes.shape({
      start: PropTypes.string,
      end: PropTypes.string,
      href: PropTypes.string,
      onClick: PropTypes.func,
    })
  ),
  interpolations: PropTypes.objectOf(
    PropTypes.oneOfType([PropTypes.element, PropTypes.string, PropTypes.number])
  ),
  tag: PropTypes.shape({
    start: PropTypes.string,
    end: PropTypes.string,
  }),
};

export default LinkText;
