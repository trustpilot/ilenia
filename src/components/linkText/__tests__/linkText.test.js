import React from 'react';
import {render, Simulate, wait} from 'react-testing-library';
import renderer from 'react-test-renderer';
import LocalizationProvider from '../../localizationProvider';
import LinkText from '../';

const translations = {
  'test1': 'No links in the string',
  'test2': 'A link [LINK-BEGIN]in the[LINK-END] middle',
  'test3': 'A link [LINK-BEGIN]in the[LINK-END] middle and one [LINK2-BEGIN]in the end[LINK2-END]',
  'test4': 'A link [LINK-BEGIN]in the[LINK-END] middle',
};

const setupLink = (string, links) => (
  <LocalizationProvider locale="en-US" translations={translations}>
    <LinkText string={string} links={links} translations={translations}/>
  </LocalizationProvider>
);

test('Renders a string with no links in it', () => {
  const component = renderer.create(setupLink('test1', []));

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Renders a string with one link in the middle', () => {
  const component = renderer.create(setupLink('test2', [{ href: 'http://something' }]));

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Renders a string with multiple links in it', () => {
  const links = [
    {
      href: 'http://something1',
    },
    {
      start: '[LINK2-BEGIN]',
      end: '[LINK2-END]',
      href: 'http://something2',
    }];

  const component = renderer.create(setupLink('test3', links));

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Renders a string with a link with arbitrary attributes', () => {
  const links = [{
    href: 'http://something1',
    target: '_blank',
    class: 'button',
  }];

  const component = renderer.create(setupLink('test4', links));

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Renders a string with a working onClick callback', () => {
  const onClickSpy = jest.fn();

  const links = [{
    href: 'http://something1',
    target: '_blank',
    class: 'button',
    onClick: onClickSpy,
  }];

  const {getByText} = render(setupLink('test4', links));
  Simulate.click(getByText('in the'));
  expect(onClickSpy.mock.calls.length).toBe(1);

  const component = renderer.create(setupLink('test4', links));
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Renders nothing when the string is not there', () => {
  console.error = jest.fn(); // eslint-disable-line no-console
  const component = renderer.create(setupLink('not-there'));

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  expect(console.error).toHaveBeenCalled(); // eslint-disable-line no-console
});
