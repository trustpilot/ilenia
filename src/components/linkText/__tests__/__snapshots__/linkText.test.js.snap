// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders a string with a link with arbitrary attributes 1`] = `
Array [
  "A link ",
  <a
    className="button"
    href="http://something1"
    target="_blank"
  >
    in the
  </a>,
  " middle",
]
`;

exports[`Renders a string with multiple links in it 1`] = `
Array [
  "A link ",
  <a
    href="http://something1"
  >
    in the
  </a>,
  " middle and one ",
  <a
    href="http://something2"
  >
    in the end
  </a>,
]
`;

exports[`Renders a string with no links in it 1`] = `"No links in the string"`;

exports[`Renders a string with one link in the middle 1`] = `
Array [
  "A link ",
  <a
    href="http://something"
  >
    in the
  </a>,
  " middle",
]
`;

exports[`Renders nothing when the string is not there 1`] = `<span />`;
